text/0000-natvis.md

```diff
@@ -0,0 +1,506 @@
+can be overridden by specifying a new manifest key.
+
+Cargo would also add new syntax in `Cargo.toml` for specifying the list of Natvis
+files to be added to the crate directly. The new manifest key, `natvis` would be
+added to a new `[debugger-visualizations]` section. Cargo would emit a warning if
+any of the specified Natvis files do not have the `.natvis` extension.
+
+As with the auto-discover feature, this would be in control of setting the
+`-C natvis` flag in rustc as well as `-Z unstable-options` to enable it. Using
+the new toml syntax to specify `.natvis` files would override Cargo from
+auto-discovering Natvis files in the `dbgvis/natvis/` directory.
+
+For example:
+
+`Cargo.toml`:
+
+```toml
+cargo-features = ["natvis"]
+
+[package]
+name = "natvis"
+version = "0.1.0"
+edition = "2018"
+
+[debugger-visualizations]
+natvis = ["a.natvis", "b.natvis"]
+```
+
+This would generate a call to rustc similar to the following,
+(for simplicity purposes, most of the rustc command line has been removed):
+
+`rustc -C natvis=path/to/file/a.natvis -C natvis=path/to/file/b.natvis -Z unstable-options`
+
+`.natvis` files that contain spaces or commas within the path will be quoted to ensure
+the path remains valid when passing command line options to rustc.
+
+The `CrateRoot` type would also need to be updated to account for `.natvis`
+files for crates within the dependency graph. To reflect this a new type
+`pub struct NatvisFile` will be created to ensure the contents of a `.natvis`
+file can be serialized and stored in the crate metadata. The `CrateRoot` would
+contain the field, `natvis_files: Lazy<[NatvisFile]>`.
+
+Another change that would need to be made here is to add a new field to the
+`CrateInfo` type, `pub natvis_files: FxHashMap<StableCrateId, Vec<NatvisFile>>`.
+This will allow the `MsvcLinker` type to query the list of Natvis files that exist
+within the crate dependency graph and add the `/NATVIS` linker argument for each
+`.natvis` file. This will also be the stage at which the Natvis files for each
+crate are copied to the target directory which they will be linked from. For example,
+in a debug build, the `.natvis` files will be copied to `target/debug/deps/natvis`.
+Each `.natvis` file that is copied over to the new directory will have a new name to
+ensure it is unique across `.natvis` files for all crates. The naming scheme will be
+`<crate_name>-<stable_crate_id>-<file_name>-<count>.natvis`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+One drawback here is that a lot of types implement the Debug trait which
+already specifies how a type should be viewed when debugging. Implementing
+this RFC would mean a Rust developer would have to manually specify the
+Natvis for a type that may already have implemented the Debug trait which
+would be redundant. Currently, running the Debug trait in the debugger directly
+is not possible and so a manual definition would be required to have a debugger view.
```

Comment:
I'm glad you mentioned this downside, because when I read the idea at the start of this RFC, it was the first thing that came to my mind. We could trivially have miri execute the MIR of a Debug impl. The only unimplemented part of this is asking the debugger for the parts of the real memory needed to run the Debug impl.

Instead of opening the road to supporting N (there will be more) debug visualization schemes (when we already derive or implement one), we could totally go down the road of adding the infrastructure for invoking miri. I'd be happy to iterate on this idea and help writing an RFC for it. I think it would be much more useful, as it would not require people to actually change their crates, we get this feature in hindsight to existing crates.

Reply:
When doing this via miri or some derivative of it we can actually be more powerful than plain visualization. We can inform about nested debug formatting runs (allowing a debugger to go from a nested visualization to its memory) and generally use the `fmt::Formatter` API to create structured output instead of just a stringification.